megamek/src/megamek/client/RandomNameGenerator.java

```diff
@@ -375,49 +268,186 @@ public int getPercentFemale() {
         return percentFemale;
     }
 
-    public void setPerentFemale(int i) {
+    public void setPercentFemale(int i) {
         percentFemale = i;
     }
 
     /**
      * randomly select gender
-     * 
+     *
      * @return true if female
      */
     public boolean isFemale() {
         return Compute.randomInt(100) < percentFemale;
     }
 
-    public static void initialize() {
-        if ((rng != null) && (rng.last != null)) {
-            return;
+    public static RandomNameGenerator getInstance() {
```

Comment:
```java
    // volatile to ensure readers get the current version
    // this is only okay because we only ever have one
    // writer and no writer performs a read-before-write
    private volatile boolean initialized;

    // synchronized ensures only one thread is entering `getInstance` at a time
    public static synchronized RandomNameGenerator getInstance() {
        // only this code reads and writes `rng`
        if (rng == null) {
            // synchronized ensures this will only be entered exactly once
            rng = new RandomNameGenerator();
            rng.runThreadLoader();
        }
        return rng;
    }

    // unchanged, it is only run once
    private void runThreadLoader() {
        Thread loader = new Thread(() -> {
            rng.populateNames();
            rng.removeInitializationListener();
        }, "Random Name Generator name initializer");
        loader.setPriority(Thread.NORM_PRIORITY - 1);
        loader.start();
    }

    private void populateNames() {
        // ... everything the same ...

        // there is only ever one invocation of us
        // so only one writer to this field is fine
        // and even if not, no writers do a
        // read-before-write.
        initialized = true;
    }
```
You would then use `initialized` in your downstream code.